### PR TITLE
PR for 134: error handling in location picker

### PR DESCRIPTION
**Pull request by lemonboston**
PR for #134 

This is based on current location branch (95), so after that is merged, this can be rebased to master.

@dmfs Please review

Known issues:
- The Google error message disappears upon rotation
- The Google error message for the current location doesn’t reappear after typing and clearing it back

I suggest creating a low priority bug to the backlog about these.

A specific note:
There are 2 new classes that at lastest state are not used: `SubClickable`: allows to decorate `ListItem` with action for a child view id. `CustomTexted`: allows to decorate `ListItem` to customize a text of a child TextView id. I used them for `ButtonedMessageItem` at a point, but it looked better eventually if this `ListItem` actually gets its required input in the constructor. But I would still keep those 2 decorators, they can come useful later.

### CODE REVIEW

**Comment by lemonboston** [comment]
Bug created for the mentioned known issues: https://github.com/schedjoules/android-event-discovery-sdk/issues/148

**Comment by dmfs** [comment]
90 files changed!? Are you sure this is just about error handling?

**Comment by dmfs** [comment]
Oh wait, that includes #95, it's actually 29 files. Still a lot though.

**Comment by lemonboston** [comment]
Yes, and it's also probably not worth reviewing until both current location and recent locations branches are merged and this is updated. So I move it back from Remove column and reassign to myself now.

**Comment by lemonboston** [comment]
Rebased this to master, but this still has to wait until #142 is merged, there will be conflicts, so I will update and review this again.

**Comment by lemonboston** [comment]
@dmfs This is ready for review.
We will need translations here as well after you reviewed.

**Comment by lemonboston** [comment]
Rebased to master, ready for review.
(We will need the translations for new strings here after the code and English versions are reviewed.)

File: eventdiscovery-sdk/src/main/res/values/strings.xml
Comment (dmfs) [inline eventdiscovery-sdk/src/main/res/values/strings.xml]: What is this message all about? Why would the user care if the screen gets reloaded or not?

File: eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/googleapiclient/BasicGoogleApiClientHolder.java
Comment (dmfs) [inline eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/googleapiclient/BasicGoogleApiClientHolder.java]: class not final

File: eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/googleapiclient/DispatchingOnConnectionFailedListener.java
Comment (dmfs) [inline eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/googleapiclient/DispatchingOnConnectionFailedListener.java]: class not final

File: eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/googleapiclient/GoogleApiClientHolder.java
Comment (dmfs) [inline eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/googleapiclient/GoogleApiClientHolder.java]: This interface looks a bit fragile too me. A builder that can be built only once? I don't like the temporal coupling here.

What we need is not a holder but a wrapper or facade to a GoogleApiClient which provides the methods we need and forwards them to a GoogleApiClient (which is built lazily one the first access). If a component needs to add another API, the wrapper will transparently recreate a new GoogleApiClient (so the classes which use the facade won't notice). This way all modules can share the facade, but you still can enable APIs as needed. Does that make sense?

File: eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/PlaceSuggestionModule.java
Comment (dmfs) [inline eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/PlaceSuggestionModule.java]: class not final

File: eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/searchlist/resultupdates/ClearNoMatterOfQuery.java
Comment (dmfs) [inline eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/searchlist/resultupdates/ClearNoMatterOfQuery.java]: I would call this `ForcedClear`, seems a bit more readable. Same for `ShowSingleNoMatterOfQuery` which I would call `ForcedShowSingle`.

**Review by dmfs — CHANGES_REQUESTED** [review]
There are quite a couple of `final` modifiers missing.

File: eventdiscovery-sdk/src/main/res/values/strings.xml
Comment (lemonboston) [reply eventdiscovery-sdk/src/main/res/values/strings.xml]: You are probably right, it's not needed to add that to the label, so I removed it.

File: eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/googleapiclient/BasicGoogleApiClientHolder.java
Comment (lemonboston) [reply eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/googleapiclient/BasicGoogleApiClientHolder.java]: Added all missing final modifiers in the project. (Will update the project file template later)

File: eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/googleapiclient/GoogleApiClientHolder.java
Comment (lemonboston) [reply eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/googleapiclient/GoogleApiClientHolder.java]: Long story short: I looked into this and started to refactor but I suggest that we create a separate technical task for this, so that this PR can be merged.

I've pushed those changes to separate branch: `stories/134-error-handling-google-api-facade'
The complexity there is with the callbacks. At this point I am still not sure if everything will work correctly with this dynamically recreated GoogleApiClients, so I would like to get back to that and review, but please have a first look at it yourself as well. The interface looks like this
```
public interface GoogleApiClientFacade
{
    void addApi(@NonNull Api<? extends Api.ApiOptions.NotRequiredOptions> googleApi);

    void addConnectionFailedListener(GoogleApiClient.OnConnectionFailedListener listener);

    void registerConnectionCallback(GoogleApiClient.ConnectionCallbacks connectionCallbacks, boolean forceCreate);

    PendingResult<AutocompletePredictionBuffer> getAutocompletePredictions(String query, AutocompleteFilter filter);

    PendingResult<PlaceBuffer> getPlaceById(String placeId);

    @RequiresPermission(anyOf = { "android.permission.ACCESS_COARSE_LOCATION", "android.permission.ACCESS_FINE_LOCATION" })
    Optional<Location> getLastLocation();
}
```

I am quite confident that the current version in this PR doesn't have risk in how this `GoogleApiClientHolder` works, the temporal coupling there is intended and used correctly now. I agree though that it's not good design like this as you pointed out, for example the client could be built with the builder directly now, which would be a different instance. But as it is used now, it should be fine, so I think merging is fine.

File: eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/searchlist/resultupdates/ClearNoMatterOfQuery.java
Comment (lemonboston) [reply eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/searchlist/resultupdates/ClearNoMatterOfQuery.java]: I like those names, too, renamed them.

**Comment by lemonboston** [comment]
Ready for review again.

**Comment by lemonboston** [comment]
Just a reminder about the currently missing translations.

**Review by dmfs — APPROVED** [review]

**Pull request merged**
